internal_worker.go

```diff
@@ -546,20 +554,59 @@ func (th *hostEnvImpl) registerEncodingTypes(fnType reflect.Type) error {
 	return nil
 }
 
-func (th *hostEnvImpl) registerValue(v interface{}) error {
-	rType := reflect.Indirect(reflect.ValueOf(v)).Type()
-	return th.registerType(rType)
+func (th *hostEnvImpl) registerValue(v interface{}, encoder encoding) error {
+	if val := reflect.ValueOf(v); val.IsValid() {
+		if val.Kind() == reflect.Ptr && val.IsNil() {
+			return nil
+		}
+		rType := reflect.Indirect(val).Type()
+		return th.registerType(rType, encoder)
+	}
+	return nil
 }
 
 // register type with our encoder.
-func (th *hostEnvImpl) registerType(t reflect.Type) error {
+func (th *hostEnvImpl) registerType(t reflect.Type, encoder encoding) error {
```

Comment:
The encoder interface was kept generic taking interface{} instead of reflection types. some of this is specific to GO, i would leave the encoding interface as it is, since registration in some serialization in advance might not make sense.